### ISSUE TITLE
Minimum Annotation Requirement 

### ISSUE THREAD

**Issue by mitras1210** [YttriLab/A-SOID#35]
I was wondering what the minimum annotation requirement is for an DLC dataset. How much annotation do we need to provide, at a minimum, for A-SOiD to be able to carry out its process efficiently?

Thank you!

**Comment by JensBlack** [comment]
Hi Mitras,

That is a great question; like many great questions, there is no simple answer to this (unfortunately).

As in many machine learning solutions, the standard practice in A-SOiD is to isolate a test set (30% of the data in our case) before starting the active learning process. This helps us verify that the algorithm learns something relevant and does not overfit the training data. To make extra sure, we even use a cross-validation step (10 splits is the standard parameter in the app).

That being said, there is a lower limit to:

A. when the active learning process works at all
B. when the test set is representative of the real-world distribution of data (your later use case)

I am unsure what you mean by "process efficiently", so let me get into both, and you can ask further if this wasn't clear enough.

> Disclaimer: Not every data set is the same, so I cannot give you a concrete number for each hypothetical use case.

## A. The active learning process

The active learning process works regardless of the number of samples always in the same way:
1. A baseline classifier is trained on the full data set (the dotted line), to give users an idea of what the classifier's performance would be with all training data available.
**This requires at least one sample per class in training as well as the test set to work at all.**
2.  An iteration-0 classifier is trained with a minimum number of samples (selected by you) to begin the automatic active learning part.
**This requires at least one sample per class to work at all.**
3. The remaining training set is used to extract low-confidence examples based on the classification of the previous iteration (in this case, iteration 0), which are added to the training data of the next iteration (in this case iteration 1).
**This requires at least one additional sample per class in the remaining training data.**

> If the remaining training data is not predicted with low confidence, no data will be added to the training set of the next iteration, and the active learning process will end.

4. This process repeats until there are:
- No more low-confidence examples
-  No additional training data is available
- The selected maximum number of iterations was reached

In general, it is a good idea to play around with the parameters and see which combination will result in the highest performance, but there are some practical tips I can give you:

a. The baseline model is better than the active learning model after the final iteration. 

This can be the case if you restrict the number of iterations too far, and the classifier would be able to perform better if you'd give it more iterations.

> Play around with the maximum iterations and maximum samples per iteration parameters.

b. The active learning classifiers reach the baseline but do not beat it.

You either have a well-balanced raw data set (good for you!), or the number of samples is too low to efficiently utilize active learning.

c. The classifier performance baseline & active learning are poor.

 If the performance level is unsatisfactory, I recommend adding new samples of the poorly performing classes. 

We are currently working on a solution to allow manual active learning within the app so that you are not forced to label entire blocks of data (e.g., new sessions) but low-confidence examples directly. So stay tuned!

## B. The real-world distribution of data (your later use case)

Be aware that with every data set you use to teach a machine learning algorithm something about your real-world use case, you are limiting its exposure to a subset of your data.

As you mentioned DLC, I'll use an analogy to pose estimation:
If you train a model to recognize the head of an animal without any attachments, implants, etc., it will become confused if presented with any animals with said additional details later on. Because we tend to stop the learning process when the model is good at predicting the test set and often do not verify its performance on the real-world data later on, it is very important to collect a ground truth data set that represents your later use case and most, if not all, relevant edge cases.

>  The smaller the data set and the test set, the more likely we will miss some relevant data.

While A-SOiD enables you to use data sets effectively (by balancing the samples and concentrating on low-confidence examples, e.g., edge cases), it does not save users from collecting initial ground data that lack representative power.

To help users build such a data set, we are now expanding our functionality.
Because we, as users ourselves, are aware that in most research cases, the available data (e.g., a recording session of an animal behaving) is inherently unbalanced - i.e., a lot of irrelevant or highly repetitive behaviors and a few relevant, low-frequency behaviors.

**Comment by koosjedr** [comment]
This requires at least one sample per class in training as well as the test set to work at all. What is meant by one sample in training and one test set. Are these both different from each other?

**Comment by JensBlack** [comment]
@koosjedr 

From the input data, A-SOiD extracts time-window sized features. E.g., with a time-window of 100 ms and a sample rate of 30 Hz, the features would be extracted from 3 frames (from the pose estimation) and 1 label (for a annotation sample rate of 10 Hz). Then, the resulting feature+label pairs are split into a train and test set. Train will be used for the model training, while test will be used to test the model. Therefore, the process needs at least 1 sample in train and 1 sample in test to be able to train/test correctly.

TLDR: They are not the same, but come from the same pool of data. They are both needed for the model to work.
